Spanning ring

krnowak opened the pull request:
This work has three areas:

- Zone negotiation.
  - Zone is supervisor's view of a network. So network inside the k8s cluster would be one zone and network outside it could be another zone.
  - Zones currently are organized in a tree-like hierarchy.
  - Normally a service in a child zone can initiate a communication with a service in parent zone, but the converse is false.
  - This might be a simplification, but I think it catches most of the networking cases.
    - This likely is not true in case of local k8s cluster created by minikube, though - the networking minikube sets up is… complicated. There are some hacks to make this scenario work, but I would like to drop them.
- Exposing supervisor (through extended member info in protocol messages) and its services (through extended service gossip messages). This basically is:
  - Watching a TOML file, where we describe how the supervisor and the services it manages are exposed.
  - Applying that information to our member information and to our service information.
    - Member information is usually sent as a part of either gossip messages or SWIM messages.
    - Service information is usually sent as generated service rumors.
- Consuming the extended member and gossip messages to properly bind services across zones.
  - This drops all the service information from census if the service is not reachable (like if it is running inside the k8s cluster and is not exposed outside).

Some things that could be done next:
- Drop address discovery.
  - While this is convenient to have, it may interfere with next point. Also, address discovery in some places looks like heuristics. Would still like to do it. This is something that probably could be done later.
- Add siblings to zones.
  - So zones are in parent-child relationships, but there could be a situation where we two k8s clusters with some supervisors exposed and there is no supervisor we can talk to outside the clusters.
- Add detection if two separate children of a parent zone are actually the same zone
  - That would likely be a thread where we send pings to supervisors in the sibling zone through their private IP to see if we get an ACK. If so, the zones are not siblings but aliases.
- Drop predecessors over time.
  - During zone negotiation we may end up with many aliases for a single zone. Would be nice to phase them out gradually to keep only the zone with the highest ID.
- Limit the number of zone aliases and zone children sent in pings (to avoid breaking the 4k bytes limit in some cases).
- Better debug story for zones code.
  - The debug was done the way it is, because I wanted to have a readable multiline debug messages from a server when running a test that has several servers running and printing messages at the same time.

Fixes #4211

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

baumanj commented:
Looks like there are some test failures. Let me know if you need any travis help there, otherwise I'll plan to take a look when those are resolved. Does that sound good, @krnowak ?

krnowak commented:
@baumanj: I updated the PR with four additional commits that should address a failure I was sometimes getting. But I still sometimes am getting failures on the `nat::nat::different_zones_get_different_ids_with_unexposed_servers` test - two servers don't see each other as healthy, but I'm not sure why, yet. Seems to be easier to trigger when other tests are running in parallel, so I don't know if I'm getting some starvation here…

Also, I'm sorry for the lousy pull request message, will update it tomorrow.

krnowak commented:
Added a bunch of commits to fix some intermittent failures I was getting. I also dropped the discovery feature from the code, so the supervisors needs to be told explicitly about its exposed address. Also rebased to current master branch.

krnowak commented:
Rebased to current master and refactored zone handling a bit. Added some unit tests for that.

baumanj commented:
I'm going to be out for a bit, so @christophermaier is going to take over reviewing this. Thanks for your patience.

baumanj commented:
This isn't under active review, so closing for now.
